<!DOCTYPE html>
<html lang="en-us">
  
  <head>
  <meta charset="UTF-8">
  <title>Genomic pipelines in Kundaje lab</title>
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="theme-color" content="#157878">
  <link rel="stylesheet" href="css/normalize.css">
  <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
  <link rel="stylesheet" href="css/cayman.css">
</head>


  <body>
    <section class="page-header">
  <h1 class="project-name">Genomic pipelines in Kundaje lab</h1>
  <h2 class="project-tagline">BigDataScript pipelines, libraries and programming guideline</h2>
  <a href="./index.html" class="btn">Overview</a>
  <a href="./usage.html" class="btn">Usage</a>
  <a href="./programming.html" class="btn">Programming</a>
  <a href="./troubleshooting.html" class="btn">Troubleshooting</a>
</section>


    <section class="main-content">
      
      <h3>/bin/bash: module: line 1: syntax error: unexpected end of file</h3>

<p>If see the following error when you submit jobs to Sun Grid Enginee,</p>
<div class="highlight"><pre><code class="language-" data-lang="">/bin/bash: module: line 1: syntax error: unexpected end of file
</code></pre></div>
<p>Check if your $HOME/.bashrc has any errorneous lines.</p>

<p>Remove the following line in you module initialization scripts ($MODULESHOME/init/bash or /etc/profile.d/modules.sh).</p>
<div class="highlight"><pre><code class="language-" data-lang="">export -f module
</code></pre></div>
<h3>Unable to run job: unknown resource &quot;&#39;mem&quot;</h3>

<p>Replace <code>$HOME/.bds/bds.config</code> with the one in the repo.</p>
<div class="highlight"><pre><code class="language-" data-lang="">$ cp /path/to/repo/bds.config $HOME/.bds/
</code></pre></div>
<h3>Unable to access jarfile /picard.jar</h3>

<p>Define a shell variable <code>PICARDROOT</code> for your environment. Add the following to your <code>~/.bashrc</code> or conda <code>activate</code>:</p>
<div class="highlight"><pre><code class="language-" data-lang="">export PICARDROOT=/path/to/your/picard-tool
</code></pre></div>
<h3>awk: cmd. line:1: fatal: division by zero attempted</h3>

<p>This error happens when 1) picard tool&#39;s MarkDuplicate is running out of memory, 2) fastq inputs have wrong endedness (SE or PE) or 3) input raw bam is incorrect.
For 1) balance memory usage among parallel tasks, add <code>-no_par</code> or reduce max. # threads (<code>-nth [SMALL_NUMBER]</code>).
For 2) check your fastq inputs are correct (<code>-fastqN_1</code>, <code>-fastqN_2</code>, ...) and also check their endedness (SE or PE) parameters like (<code>-se</code> or <code>-pe</code>).
For 3) check if there is an error in aligning stage (in an HTML report).</p>

<h3>Unsupported major.minor version (java)</h3>

<p>When running bds (BigDataScript), you get the following error if you have lower version of java or high version of java is not selected as default.</p>

<p>Solution:</p>
<div class="highlight"><pre><code class="language-" data-lang="">$ sudo apt-get install openjdk-8-jre # for Fedora based linux (Red Hat, ...)

$ sudo yum install java-1.8.0-openjdk # for Debian based linux (Ubuntu, ...)

$ sudo update-alternatives --config java # choose the latest java as default
</code></pre></div>
<h3>[main_samview] random alignment retrieval only works for indexed BAM or CRAM files.</h3>

<p>If your pipeline starts from BAM files, make sure that bam index (.bam.bai) exists together with BAM file. If not, build index with <code>samtools index [YOUR_BAM_FILE]</code>. BAM and BAI should be in the same directory.</p>

<h3>Fatal error: /home/leepc12/bds<em>atac/modules/report</em>*.bds</h3>

<p>Simply re-run the pipeline with the same command. Possible bug in BDS for locking/unlocking global variables.</p>

<h3>ImportError: libopenblasp-r0-39a31c03.2.18.so: cannot open shared object file: No such file or directory</h3>

<p>Dependencies are not installed correctly. Check your Anaconda Python is correctly configured for conda environments. Run <code>./uninstall_dependencies.sh</code> and then <code>./install_dependencies.sh</code> again.</p>

<h3>Unable to run job: unknown resource &quot;mem&quot;</h3>

<p>Copy <code>./bds.config</code> to <code>$HOME/.bds/</code>.</p>

<h3>Error trying to find out post-mortem info on task</h3>

<p>For fast scheduling clusters including SGE, doing post-mortem on jobs can fail in BDS. Add <code>clusterPostMortemDisabled = true</code> to your <code>~/.bds/bds.config</code>.</p>

<h3>java.lang.OutOfMemoryError: unable to create new native thread</h3>

<p>Number of threads created by BDS exceeds limit (<code>ulimit -a</code>). BDS created lots of thread per pipeline (more than 20). So if you see any thread related error, check your <code>ulimit -a</code> and increase it a bit.</p>

<h3>Task disappeared</h3>

<p>Check a log of failed tasks with <code>qacct -j [JOB_ID]</code> (for SGE). You job can be killed due to resource settings. Increase your walltime or max. memory (<code>-wt</code>, <code>-wt_APPNAME</code>, <code>-mem</code> or <code>-mem_APPNAME</code>) for the task of failure.</p>

<h3>File exists, No file or directory (related to parallel conda activations)</h3>

<p>This is a known bug in conda <a href="https://github.com/conda/conda/issues/2837">#2837</a> and has not been fixed yet even in the latest version (4.2.1). Downgrade conda to 4.0.5 or 4.0.10. </p>
<div class="highlight"><pre><code class="language-" data-lang="">Traceback (most recent call last):
  File "/home/leepc12/miniconda3/bin/conda", line 6, in &lt;module&gt;
    sys.exit(main())
  File "/home/leepc12/miniconda3/lib/python3.5/site-packages/conda/cli/main.py", line 48, in main
    activate.main()
  File "/home/leepc12/miniconda3/lib/python3.5/site-packages/conda/cli/activate.py", line 135, in main
    conda.install.symlink_conda(prefix, root_dir, shell)
  File "/home/leepc12/miniconda3/lib/python3.5/site-packages/conda/install.py", line 596, in symlink_conda
    symlink_conda_hlp(prefix, root_dir, where, symlink_fn)
  File "/home/leepc12/miniconda3/lib/python3.5/site-packages/conda/install.py", line 613, in symlink_conda_hlp
    symlink_fn(root_file, prefix_file)
FileExistsError: [Errno 17] File exists: '/home/leepc12/miniconda3/bin/conda' -&gt; '/home/leepc12/miniconda3/envs/bds_atac/bin/conda'

Traceback (most recent call last):
  File "/home/leepc12/miniconda3/bin/conda", line 6, in &lt;module&gt;
    sys.exit(main())
  File "/home/leepc12/miniconda3/lib/python3.5/site-packages/conda/cli/main.py", line 48, in main
    activate.main()
  File "/home/leepc12/miniconda3/lib/python3.5/site-packages/conda/cli/activate.py", line 135, in main
    conda.install.symlink_conda(prefix, root_dir, shell)
  File "/home/leepc12/miniconda3/lib/python3.5/site-packages/conda/install.py", line 596, in symlink_conda
    symlink_conda_hlp(prefix, root_dir, where, symlink_fn)
  File "/home/leepc12/miniconda3/lib/python3.5/site-packages/conda/install.py", line 610, in symlink_conda_hlp
    os.remove(prefix_file)
FileNotFoundError: [Errno 2] No such file or directory: '/home/leepc12/miniconda3/envs/bds_atac_py3/bin/conda'
</code></pre></div>
<h3>Exception in thread &quot;main&quot; java.lang.NumberFormatException: For input string: &quot;40G&quot;</h3>

<p>Do not use <code>-mem</code> in your command line. Use &#39;-memory` instead.</p>
<div class="highlight"><pre><code class="language-" data-lang="">$ bds chipseq.bds -mem 40G
Picked up _JAVA_OPTIONS: -Xms256M -Xmx1024M -XX:ParallelGCThreads=1
Exception in thread "main" java.lang.NumberFormatException: For input string: "40G"
        at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
        at java.lang.Long.parseLong(Long.java:589)
        at java.lang.Long.parseLong(Long.java:631)
        at org.bds.lang.Type.parse(Type.java:334)
        at org.bds.BdsParseArgs.parseArgs(BdsParseArgs.java:207)
        at org.bds.BdsParseArgs.initializeArgs(BdsParseArgs.java:150)
        at org.bds.BdsParseArgs.initializeArgs(BdsParseArgs.java:106)
        at org.bds.BdsParseArgs.parse(BdsParseArgs.java:172)
        at org.bds.Bds.runCompile(Bds.java:872)
        at org.bds.Bds.run(Bds.java:815)
        at org.bds.Bds.main(Bds.java:182)
</code></pre></div>

      <footer class="site-footer">
  <span class="site-footer-owner"><a href="">Genomic pipelines in Kundaje lab</a> is maintained by <a href="https://sites.google.com/site/anshulkundaje/Home">Jin Lee and Anshul Kundaje</a>.</span>
  <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/pietromenna/jekyll-cayman-theme">Cayman theme</a> by <a href="http://github.com/jasonlong">Jason Long</a>.</span>
</footer>


    </section>

  </body>
</html>
